cluster/addons/dns/kube2sky/kube2sky.go

```diff
@@ -441,34 +480,20 @@ func (ks *kube2sky) updateService(oldObj, newObj interface{}) {
 	ks.newService(newObj)
 }
 
-func newEtcdClient(etcdServer string) (*etcd.Client, error) {
-	var (
-		client *etcd.Client
-		err    error
-	)
-	for attempt := 1; attempt <= maxConnectAttempts; attempt++ {
-		if _, err = etcdutil.GetEtcdVersion(etcdServer); err == nil {
-			break
-		}
-		if attempt == maxConnectAttempts {
-			break
-		}
-		glog.Infof("[Attempt: %d] Attempting access to etcd after 5 second sleep", attempt)
-		time.Sleep(5 * time.Second)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("failed to connect to etcd server: %v, error: %v", etcdServer, err)
-	}
-	glog.Infof("Etcd server found: %v", etcdServer)
-
+func newEtcdClient(config *etcdstorage.EtcdConfig) (client etcd.Client, err error) {
+	ctx := context.TODO()
 	// loop until we have > 0 machines && machines[0] != ""
-	poll, timeout := 1*time.Second, 10*time.Second
+	poll, timeout := pollInterval, pollTimeout
 	if err := wait.Poll(poll, timeout, func() (bool, error) {
-		if client = etcd.NewClient([]string{etcdServer}); client == nil {
-			return false, fmt.Errorf("etcd.NewClient returned nil")
+		if client, err = config.NewEtcdClient(); err != nil {
+			glog.Infof("cannot create etcd client: %v", err)
+			return false, nil
+		}
+		if err = client.Sync(ctx); err != nil {
```

Comment:
If this exits, kubelet will just start it again and the rest of the
kubernetes control plane will see the fact that dns is having trouble
starting, so I think exiting is OK.

On Thu, Mar 24, 2016 at 6:15 AM, a-palchikov <notifications@github.com>
wrote:

> In cluster/addons/dns/kube2sky/kube2sky.go
> <https://github.com/kubernetes/kubernetes/pull/23331#discussion_r57313980>
> :
>
> >  	if err := wait.Poll(poll, timeout, func() (bool, error) {
> > -		if client = etcd.NewClient([]string{etcdServer}); client == nil {
> > -			return false, fmt.Errorf("etcd.NewClient returned nil")
> > +		if client, err = config.NewEtcdClient(); err != nil {
> > +			glog.Infof("cannot create etcd client: %v", err)
> > +			return false, nil
> > +		}
> > +		if err = client.Sync(ctx); err != nil {
>
> I agree that the explicit connectivity verification (as it is here) is
> probably unnecessary.
>
> On a side note, I'm a bit concerned about the mutation timeout - if etcd
> is inaccessible within mutation timeframe (10s by default), kube2sky
> breaks. Is this a reasonable requirement? Can it not just keep retrying
> forever?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23331/files/05ae2d0cfda496dd91e181855e55a0b5945674c0#r57313980>
>